rr_graph_uxsdcxx.cpp

```diff
@@ -0,0 +1,3174 @@
+#include "rr_graph_uxsdcxx.h"
+
+/*
+ * This file is generated by uxsdcxx.
+ * Modify only if your build process doesn't involve regenerating this file.
+ *
+ * Cmdline: ./uxsdcxx.py rr_graph.xsd
+ * Input file: rr_graph.xsd
+ * md5sum of input file: 2b7bc77a1e7c006250cf0351f8046f9e
+ */
+
+/* All uxsdcxx functions and structs live in this namespace. */
+namespace uxsd {
+
+std::vector<t_pin> pin_pool;
+std::vector<t_meta> meta_pool;
+std::vector<t_pin_class> pin_class_pool;
+std::vector<t_x_list> x_list_pool;
+std::vector<t_y_list> y_list_pool;
+std::vector<t_switch> switch_pool;
+std::vector<t_segment> segment_pool;
+std::vector<t_block_type> block_type_pool;
+std::vector<t_grid_loc> grid_loc_pool;
+std::vector<t_node> node_pool;
+std::vector<t_edge> edge_pool;
+std::vector<t_connection_box> connection_box_pool;
+
+string_pool_impl string_pool;
+
+
+void clear_pools(void){
+	pin_pool.clear();
+	meta_pool.clear();
+	pin_class_pool.clear();
+	x_list_pool.clear();
+	y_list_pool.clear();
+	switch_pool.clear();
+	segment_pool.clear();
+	block_type_pool.clear();
+	grid_loc_pool.clear();
+	node_pool.clear();
+	edge_pool.clear();
+	connection_box_pool.clear();
+}
+
+void clear_string_pool(void){
+	string_pool.clear();
+}
+
+
+/* Lookup tables for enums. */
+const char *lookup_switch_type[] = {"UXSD_INVALID", "mux", "tristate", "pass_gate", "short", "buffer"};
+const char *lookup_pin_type[] = {"UXSD_INVALID", "OPEN", "OUTPUT", "INPUT"};
+const char *lookup_loc_side[] = {"UXSD_INVALID", "LEFT", "RIGHT", "TOP", "BOTTOM"};
+const char *lookup_node_type[] = {"UXSD_INVALID", "CHANX", "CHANY", "SOURCE", "SINK", "OPIN", "IPIN"};
+const char *lookup_node_direction[] = {"UXSD_INVALID", "INC_DIR", "DEC_DIR", "BI_DIR"};
+
+
+pugi::xml_parse_result rr_graph::load(std::istream &is){
+	pugi::xml_document doc;
+	pugi::xml_parse_result result = doc.load(is);
+	if(!result) return result;
+	for(pugi::xml_node node= doc.first_child(); node; node = node.next_sibling()){
+		if(std::strcmp(node.name(), "rr_graph") == 0){
+			/* If errno is set up to this point, it messes with strtol errno checking. */
+			errno = 0;
+			load_rr_graph(node, this);
+		}
+		else throw std::runtime_error("Invalid root-level element " + std::string(node.name()));
+	}
+	return result;
+}
+
+void rr_graph::write(std::ostream &os){
+	/* Print floating points with max double precision. */
+	os.precision(std::numeric_limits<double>::max_digits10);
```

Comment:
Should this be controllable in some way?

Reply:
Probably, but I couldn't think of a proper API. Maybe the caller should set this?